README.md
# PlagScan Moodle Plugin 3.2.2 Release Notes

The advanced plagiarism checker plugin for moodle.

-------------

System requirements:
--------------------

    - moodle >= 3.1
    - PlagScan organisation account

Installation instruction and manuals:
-------------------------

https://www.plagscan.com/en/plagscan-integration-moodle

Bug Fix:
--------

- Remove deprecated config variables
- Remove deprecated methods
- Added methods to support Course module edit form Callbacks
- Added support for latest Moodle versions 3.9, 3.10 and 3.11 
- Fix problem with showing some debug message from non set variables or objects

Special thanks:
---------------

    - Jesus and Kirupa for the time dedicate

Rubén Olmedo © PlagScan.com

Jesús Prieto © PlagScan.com


CHANGELOG.md
### Version: 3.2.2 Build: 2021110302

- Added new feature to submit all files from an assigment
- Fix a bug when trying to use AMOS to translate items
- Fix a bug where 'Save and display' option when editing an assignment was returning back to course instead of the assignment view
- Fix a bug where plugin admin settings page was still accessible when the plagiarism plugin option was disabled under Site Administration / Advanced features

### Version: 3.2.1 Build: 2021060900

- Remove deprecated config variables
- Remove deprecated methods
- Added methods to support Course module edit form Callbacks
- Added support for latest Moodle versions 3.9, 3.10 and 3.11 
- Fix problem with showing some debug message from non set variables or objects

### Version: 3.1.29 Build: 2020111901

- Add proxy support, getting proxy config from Moodle settings
- Fix a problem causing that some content was shown when creating/editing assigment related to exclude functions

### Version: 3.1.28 Build: 2020071501

- Add Backup/Restore functionality, allowing to backup assignments where PlagScan plugin is enabled and create new ones from there having PlagScan enabled once is restored.


### Version: 3.1.27 Build: 2020060201

- Fix issue on assignment page load due to call to a function get_pathnamehash() on null.


### Version: 3.1.26 Build: 2020040601

- Updated Scheduled Task for checking files on assignments configured with the option for checking automatically to avoid retry on files that returned in error due to run out of credits.

### Version: 3.1.25 Build: 2020040101

- Fixed bug causing permissions error when accessing to a report for involved users in an assignment

### Version: 3.1.24 Build: 2020031702

- Added defaults PlagScan settings for Assignments on the admin settings
- Added option to upload to PlagScan after final submission
  - Changed option for enable PlagScan for files submission: "Start check manually" for "Start check manually (Upload to PlagScan when students save a draft)"
  - Added new option for enable PlagScan for files submission: "Start check manually (Upload to PlagScan when students click on Submit button)"
- Fixed problem were report level class (icon color) was always red

### Version: 3.1.23 Build: 2020011000

- Add a button on admin settings to wipe the PlagScan User Cache table. This could be used to fix some inconsistency
- Add more checks and info messages in case there was some inconsistency between the Moodle Plugin data and PlagScan
- Add a retry when creates user in PlagScan in case it fails once
- Add and update strings

### Version: 3.1.22 Build: 2019110500

- Improve Deutsch translation strings
- Improve ajax calls when updating the file status
- Fix a bug that was showing plaintext data from ajax call in the interface output instead of returning to the previous page or refreshing it properly
- Improve the file status update, changing the status of the file in case of problems found during the conversion/check process
- Fix a bug that was showing the disclosure of the PlagScan Plugin even in other Moodle Activies that are not assignments, and also in assignments where PlagScan was not activated
- Improve the redirect from action to return to previous page with proper pagination

### Version: 3.1.21 Build: 2019101700

- Use trim to avoid problem setting the PlagScan server URL when adding extra trailing slash
- Remove log output that were shown during the resubmission of files

### Version: 3.1.20 Build: 2019101501

- Improve submit file process to avoid problems from assignments created with the old plugin version (<2.x.x)

### Version: 3.1.19 Build: 2019101005

- Improve submit file process to avoid problems when the files are associated to the Admin account

### Version: 3.1.18 Build: 2019101000

- Improve scheduled tasks in general and identify more problems in the process, updating the status of the file
- Add more status for files that failed during submission
- Update file status of frozen files

### Version: 3.1.17 Build: 2019100904

- Add scheduled task to resubmit files on resubmission state

### Version: 3.1.16 Build: 2019100902

- Add scheduled task to update status of files got frozen during checking progress
- Fix a problem when trying to refresh the status of the file it was not receiving the information in case the file was broken or rejected by PlagScan
- Fix support for files with extension on uppercase

### Version: 3.1.15 Build: 2019092503

- Add error message for files failed due to an error in the creation of the user or the user already was existing within another institution
- Add error message for files that failed during the status refresh of the file if the Client ID was changed on plugin settings
- Add status update for files that failed during the status refresh due to not have a plagiarism level result
- Add scheduled task to check frozen files from assignment with the option to check on deadline
- Add scheduled task to check frozen files from assignment with the option to check automatically

### Version: 3.1.14 Build: 2019090501

- Add resubmit button for file submission that were not able to be sent to PlagScan due an error (including API/server unreachable)
- Improve the submission to PlagScan workflow to detect errors during the process and show it afterwards
- Improve the assignment creation in PlagScan workflow to detec errors during the process and show it afterwards
- Added new error control and event for error creating the user in PlagScan

### Version: 3.1.13 Build: 2019090304

- Add option inside the assignment settings to enable PlagScan for files submission: submit files manually, with this option the files are send manually and not automatically by the plugin
- Change lang strings to update options

### Version: 3.1.12 Build: 2019082805

- Remove the assignment exclude options in the edit menu for assignments created from the old versions of the plugin (2.x)

### Version: 3.1.11 Build: 2019082804


db/upgrade.php
<?php

// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * database upgrade code for plagscan plugin
 *
 * @package      plagiarism_plagscan
 * @subpackage   plagiarism
 * @author       Daniel Gockel <dgockel@plagscan.com>
 * @copyright    @2015 PlagScan GmbH
 * @license      http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
function xmldb_plagiarism_plagscan_upgrade($oldversion) {
    global $DB;

    $dbman = $DB->get_manager();

    if ($oldversion < 2012031201) {
        // Define field value to be dropped from plagscan_config
        $table = new xmldb_table('plagscan_config');
        $field = new xmldb_field('assignmentid');

        // Conditionally launch drop field value
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012031201, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012031204) {
        $table = new xmldb_table('plagscan');

        // Remove the 'assignmentid' field
        $field = new xmldb_field('assignmentid');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // Remove the 'fileid' field
        $field = new xmldb_field('fileid');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // Remove the 'courseid' field
        $field = new xmldb_field('courseid');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // Remove the 'file' field
        $field = new xmldb_field('file');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // Create a new 'cmid' field
        $field = new xmldb_field('cmid', XMLDB_TYPE_INTEGER, '10', XMLDB_UNSIGNED, null, null, null, 'status');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // Create a new 'filehash' field
        $field = new xmldb_field('filehash', XMLDB_TYPE_CHAR, '255', null, null, null, null, 'cmid');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012031204, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012031300) {
        $table = new xmldb_table('plagscan_config');

        // Remove the 'assignmentid' field
        $field = new xmldb_field('name');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // Remove the 'fileid' field
        $field = new xmldb_field('value');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // Add the 'upload' field (for when to do an upload)
        $field = new xmldb_field('upload', XMLDB_TYPE_INTEGER, '4', null, XMLDB_NOTNULL, null, '0', 'cm');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // Add the 'complete' field (for when an automatic upload has been done)
        $field = new xmldb_field('complete', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNUL, null, '0', 'upload');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012031300, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012031400) {
        $table = new xmldb_table('plagscan');

        // Change 'pstatus' into a number
        $field = new xmldb_field('pstatus', XMLDB_TYPE_NUMBER, '4, 1', null, null, null, null, 'pid');
        $dbman->change_field_type($table, $field);

        // Remove the unused 'preport' field
        $field = new xmldb_field('preport');
        if ($dbman->field_exists($table, $field)) {
            $dbman->drop_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012031400, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012031401) {
        // Rename 'plagscan' table to 'plagiarism_plagscan' (to meet Moodle guidelines)
        $table = new xmldb_table('plagscan');
        $dbman->rename_table($table, 'plagiarism_plagscan');

        // Rename 'plagscan_config' table to 'plagiarism_plagscan_config' (to meet Moodle guidelines)
        $table = new xmldb_table('plagscan_config');
        $dbman->rename_table($table, 'plagiarism_plagscan_config');

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012031401, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012031402) {
        // Add the 'updatestatus' field
        $table = new xmldb_table('plagiarism_plagscan');
        $field = new xmldb_field('updatestatus', XMLDB_TYPE_INTEGER, '2', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'filehash');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012031402, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012060600) {
        // Add the 'username' field to the plagscan_config table
        $table = new xmldb_table('plagiarism_plagscan_config');
        $field = new xmldb_field('username', XMLDB_TYPE_CHAR, '100', null, null, null, null, 'complete');

        // Conditionally launch add field username
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012060600, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2012062501) {
        // Make the pid index non-unique (as mutliple records might be '0')
        $table = new xmldb_table('plagiarism_plagscan');
        $index = new xmldb_index('pid', XMLDB_INDEX_NOTUNIQUE, array('pid'));

        // Conditionally launch add index pid
        if ($dbman->index_exists($table, $index)) {
            $dbman->drop_index($table, $index);
        }
        $dbman->add_index($table, $index);

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2012062501, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2015110700) {
        //add nondisclosure field to plagiarism_plagscan_config
        $table = new xmldb_table('plagiarism_plagscan_config');
        $field = new xmldb_field('nondisclosure', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'username');

        // Conditionally launch add field nondisclosure
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }


        $field = new xmldb_field('show_report_to_students', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'nondisclosure');

        // Conditionally launch add field nondisclosure
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2015110700, 'plagiarism', 'plagscan');
    }


    if ($oldversion < 2016052501) {
        //add groups_capabilities field to plagiarism_plagscan_config
        $table = new xmldb_table('plagiarism_plagscan_config');
        $field = new xmldb_field('show_students_links', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'show_report_to_students');


        // Conditionally launch add field groups
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }


        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2016052501, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2016052502) {

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2016052502, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2016061401) {

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2016061401, 'plagiarism', 'plagscan');
    }
    if ($oldversion < 2017120601) {

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2017120601, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2018072301) {

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2018072301, 'plagiarism', 'plagscan');
    }


    if ($oldversion < 2018112000) {

        //add groups_capabilities field to plagiarism_plagscan_config
        $table = new xmldb_table('plagiarism_plagscan');
        $field = new xmldb_field('submissiontype', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'updatestatus');

        // Conditionally launch add field groups
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        $table = new xmldb_table('plagiarism_plagscan_config');
        $field = new xmldb_field('ownerid', XMLDB_TYPE_CHAR, '255', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, 'show_students_links');

        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        $field = new xmldb_field('submissionid', XMLDB_TYPE_CHAR, '255', XMLDB_UNSIGNED, null, null, null);
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        $field = new xmldb_field('enable_online_text', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'submissionid');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }

        $table = new xmldb_table('plagiarism_plagscan_user');

        $table->add_field('id', XMLDB_TYPE_INTEGER, '10', XMLDB_UNSIGNED, XMLDB_NOTNULL, XMLDB_SEQUENCE, null);

        $table->add_field('userid', XMLDB_TYPE_INTEGER, '10', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, null, 'id');

        $table->add_field('psuserid', XMLDB_TYPE_INTEGER, '10', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, null, 'userid');

        $table->add_key('primary', XMLDB_KEY_PRIMARY, array('id'), null, null);
        $table->add_key('userid', XMLDB_KEY_FOREIGN, array('userid'), 'user', array('id'));

        if (!$dbman->table_exists($table)) {
            $dbman->create_table($table);
        }

        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2018112000, 'plagiarism', 'plagscan');
    }
    
    if ($oldversion < 2019082201) {
        $table = new xmldb_table('plagiarism_plagscan_config');
        $DB->execute('UPDATE {plagiarism_plagscan_config} SET submissionid = ? WHERE submissionid = ?', 
            [NULL, 'ownerid']);
        
        $field = new xmldb_field('ownerid', XMLDB_TYPE_CHAR, '255', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, 0);
        if ($dbman->field_exists($table, $field)) {
            $dbman->change_field_default($table, $field);
        }

        $field = new xmldb_field('submissionid', XMLDB_TYPE_CHAR, '255', XMLDB_UNSIGNED, null, null, null);
        if ($dbman->field_exists($table, $field)) {
            $dbman->change_field_default($table, $field);
        }
        
        // plagscan savepoint reached
        upgrade_plugin_savepoint(true, 2019082201, 'plagiarism', 'plagscan');
    }
    
    if ($oldversion < 2019082804) {
        $table = new xmldb_table('plagiarism_plagscan_config');
        
        $field = new xmldb_field('exclude_from_repository', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'enable_online_text');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }
        
        $field = new xmldb_field('exclude_self_matches', XMLDB_TYPE_INTEGER, '1', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'exclude_from_repository');
        if (!$dbman->field_exists($table, $field)) {
            $dbman->add_field($table, $field);
        }
        
        upgrade_plugin_savepoint(true, 2019082804, 'plagiarism', 'plagscan');
    }
    
    if ($oldversion < 2019101000) {
        $DB->execute('UPDATE {plagiarism_plagscan} SET status = ? WHERE (status = ? OR status = ? OR status = ?) AND pid = ?', 
            [2000, 1000, 1002, 1004, 0]);
        
        upgrade_plugin_savepoint(true, 2019101000, 'plagiarism', 'plagscan');
    }

    if ($oldversion < 2021060900) {
        $enabled = get_config('plagiarism', 'plagscan_use');
        set_config('enabled', $enabled, 'plagiarism_plagscan');
        $DB->delete_records('config_plugins', array('name' => 'plagscan_use', 'plugin' => 'plagiarism'));
        $DB->delete_records('config_plugins', array('name' => 'plagscan_use', 'plugin' => 'plagiarism_plagscan'));
        
        upgrade_plugin_savepoint(true, 2021060900, 'plagiarism', 'plagscan');
    }

    return true;
}
